src/vm.c

```diff
@@ -374,7 +374,45 @@ fiber_terminate(mrb_state *mrb, struct mrb_context *c, mrb_callinfo *ci)
 {
   mrb_assert(c != mrb->root_c);
 
+  struct REnv *env = CI_ENV(ci);
+  mrb_assert(env == NULL || MRB_ENV_LEN(env) <= c->stend - ci->stack);
+
   c->status = MRB_FIBER_TERMINATED;
+  mrb_free(mrb, c->cibase);
+  c->cibase = c->ciend = c->ci = NULL;
+  mrb_value *stack = c->stbase;
+  c->stbase = c->stend = NULL;
+
+  if (!env) {
+    mrb_free(mrb, stack);
+  }
+  else {
+    env->cxt = NULL;
+
+    size_t len = (size_t)MRB_ENV_LEN(env);
+    if (len == 0) {
+      env->stack = NULL;
+      MRB_ENV_CLOSE(env);
+      mrb_free(mrb, stack);
+    }
+    else {
+      mrb_assert(stack == env->stack);
+      mrb_write_barrier(mrb, (struct RBasic*)env);
```

Comment:
Do we need `mrb_write_barrier()` here?

Reply:
Thanks for your review.

I am not sure exactly.
This `fiber_terminate()` is based on `mrb_env_unshare()`.

It was removed from `mrb_env_unshare()` once, but was reintroduced in commit 93648fc954c0133acc6e89860ba49f78ca60460f, so I added it.

Reply:
Oh, I see. I investigated and found out we need it. Thank you.